src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/ChainPal.Apple.cs

```diff
@@ -305,6 +308,57 @@ private static (X509Certificate2, int)[] ParseResults(
             return elements;
         }
 
+        private static void AddUnknownNameConstraints((X509Certificate2, int)[] elements)
+        {
+            // Apple as of macOS 15 does not do anything when a certificate contains a name constraint on UPNs. Rather
+            // than silently ignore them, add a flag indicating its not supported.
+            for (int i = 0; i < elements.Length; i++)
+            {
+                (X509Certificate2 cert, int flags) = elements[i];
+                X509Extension? ext = cert.Extensions[Oids.NameConstraints];
+
+                if (ext is null || ext.RawData is null)
+                {
+                    continue;
+                }
+
+                NameConstraintsAsn asn;
+
+                try
+                {
+                    asn = NameConstraintsAsn.Decode(ext.RawData, AsnEncodingRules.DER);
+                }
+                catch (CryptographicException)
+                {
+                    elements[i] = (cert, flags | (int)X509ChainStatusFlags.InvalidExtension);
+                    return;
+                }
+
+                if (AnyUpnConstraint(asn.excludedSubtrees) || AnyUpnConstraint(asn.permittedSubtrees))
+                {
+                    elements[i] = (cert, flags | (int)X509ChainStatusFlags.HasNotSupportedNameConstraint);
+                }
+            }
+
+            static bool AnyUpnConstraint(GeneralSubtreeAsn[]? subtreeAsns)
```

Comment:
Should we be opting UPN in as unknown, or say "they only handle 4 kinds, we'll reject anything else"?